### ISSUE TITLE
Signals don't trigger change detection using native async/await (ES2017+)

### ISSUE THREAD

**Issue by brentyates** [angular/angular#52940]
### Which @angular/* package(s) are the source of the bug?

Don't known / other

### Is this a regression?

No

### Description

See the stackblitz for replication.

When using ES2017+, the native async/await version, setting a signal after an await statement does not cause change detection to run. I assume that the component is marked for check, but nothing triggers a change detection cycle.

However, using the async pipe and observables in the same scenario does cause change detection to run. In fact, if you look at the stackblitz, change detection runs properly and the expected behaviour occurs just by creating a new component property using toObservable of the signal, without even using it in the template or anywhere else.

Setting the target to ES2015 also causes the expected behaviour to occur. As does using .then instead of async/await, and as does setting the signal inside of a zone.run callback.

I understand the older open issues that zone.js cannot patch the native async/await. However, signals are described to be able to trigger change detection without the need for zone.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-vetml3?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true
Did not see change detection run.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
{
  "name": "stackblitz-starters-vetml3",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "rxjs": "7.8.1",
    "tslib": "2.6.2",
    "zone.js": "0.14.2",
    "@angular/core": "17.0.2",
    "@angular/forms": "17.0.2",
    "@types/jasmine": "4.3.6",
    "@angular/common": "17.0.2",
    "@angular/router": "17.0.2",
    "@angular/compiler": "17.0.2",
    "@angular/animations": "17.0.2",
    "@angular/platform-browser": "17.0.2"
  },
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.1100.4",
    "@angular/cli": "~11.0.4",
    "@angular/compiler-cli": "~11.0.4",
    "@types/jasmine": "~3.6.0",
    "@types/node": "^12.11.1",
    "codelyzer": "^6.0.0",
    "jasmine-core": "~3.6.0",
    "jasmine-spec-reporter": "~5.0.0",
    "karma": "~5.1.0",
    "karma-chrome-launcher": "~3.1.0",
    "karma-coverage": "~2.0.3",
    "karma-jasmine": "~4.0.0",
    "karma-jasmine-html-reporter": "^1.5.0",
    "protractor": "~7.0.0",
    "ts-node": "~8.3.0",
    "tslint": "~6.1.0",
    "typescript": "~4.0.2"
  }
}
```

### Anything else?

Related: https://github.com/angular/angular/issues/31730

**Comment by JeanMeche** [comment]
This looks like a stackblitz issue. After the `await`, the code is run outside the angular zone. I can't reproduce it locally.

**Comment by brentyates** [comment]
I have a local reproduction which drove me to create the stackblitz reproduction. My local reproduction was using this:

```
Angular CLI: 17.0.0
Node: 18.15.0
Package Manager: npm 9.5.0
OS: darwin arm64

Angular: 17.0.2
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.0
@angular-devkit/build-angular   17.0.0
@angular-devkit/core            16.2.10
@angular-devkit/schematics      16.2.10
@angular/cli                    17.0.0
@angular/fire                   16.0.0
@schematics/angular             16.2.10
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2```

What version were you using @JeanMeche ?

**Comment by brentyates** [comment]
Interesting.... I re-instaleld Angular CLI globally to the latest version, generated a new app to make the repo, and I am not reproducing the problem locally. I need to look very closely at the CLI and devkit versions of my other repo, as well as the stackblitz version....

**Comment by JoostK** [comment]
This is sort of expected, although unfortunate.

Native async/await is not visible to zone.js, so any activity in those async stack frames does not automatically trigger change detection (signals, at least today, only manage reactive dependencies, they do not initiate change detection scheduling).

Since there isn't any other activity going on the page, change detection is just never scheduled again.

**Comment by JoostK** [comment]
Note that this is different from what you _observe_ using RxJS (pun intended), because the `delay`'s timer _is_ tracked by zone.js, and the corresponding subscription logic is then executed in synchronously (during the timer event) in the Angular zone, therefore initiating change detection to be scheduled.

The native `await` results in an asynchronous boundary between the `delay` timer firing and the execution of the code after `await`, as `await` _always_ introduces an extra microtick, thereby falling off zone.js's task execution stack.

**Comment by brentyates** [comment]
@JoostK I agree with you conceptually and your theory, but I don't think that quite explains what is happening here, especially since both @JeanMeche and I both have local versions where the bug does not appear. So I now have one local not working, one working, and one stackblitz not working...

I suspect it has something to do with the babel plugin that Jean pointed out, and I'm currently looking for the differences.

**Comment by JoostK** [comment]
My comments are in a world with native async/await. I suspect the differences may occur if, for some reason, async/await is being downleveled to use generators and the `Promise` constructor, as then zone.js would have been able to monkey-patch that `Promise` constructor.

**Comment by brentyates** [comment]
When I download the stackblitz project, as well as when I am using my local repo that has the issue, I have to use legacy peer deps or force to run npm install. The local repo that is working did not need force or legacy peer deps. 

As well, it does not look like my local code (both working version and non working version) are using native async/await as I had thought, but rather a generated async function:

```
var __async = (__this,__arguments,generator)=>{
    return new Promise((resolve,reject)=>{
        var fulfilled = (value)=>{
            try {
                step(generator.next(value));
            } catch (e) {
                reject(e);
            }
        }
        ;
        var rejected = (value)=>{
            try {
                step(generator.throw(value));
            } catch (e) {
                reject(e);
            }
        }
        ;
        var step = (x)=>x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
        step((generator = generator.apply(__this, __arguments)).next());
    }
    );
}
;
```

Zone.js is the same version (0.14.2) in my package-lock.json. I'm continuing to look for differences. Calling this a native async/await problem appears to be misleading, however targeting ES2015 does fix the problem... so it's still some sort of async/await related problem.

**Comment by JeanMeche** [comment]
When targeting ES2015, typescript converts to async/await (because they're a ES2017 feature). For greater versions we use the plugin.

**Comment by JeanMeche** [comment]
Ok yes, that's where the issue comes from on stackblitz. For the [v17 update](https://github.com/stackblitz/starters/pull/21), target changed from ES2015 to ES2022.

**Comment by brentyates** [comment]
@JeanMeche there must be another way to fix the stackblitz without changing the target to ES2015 thought, right? Since my local version using `ng new` is targeting ES2022 and it's working, and I'm guessing your local version that could not repro the issue was also targeting ES2022?

I'm not very familiar with how the stackblitz is running, if the stackblitz is not running `ng serve` then maybe that explains why the stackblitz cannot be fixed.

**Comment by JeanMeche** [comment]
It is very likely that stackblitz uses a custom script that doesn't rely on the babel plugin which would explain the difference. 

I've opened stackblitz/starters#24 to revert the target change.

**Comment by brentyates** [comment]
I'm satisfied with the explanations here, thanks for the help in understanding. Is this summary right @JoostK: that in the current situation signals help ensure that change detection, when run, is optimized, rather than ensuring that change detection occurs, which Angular still relies upon zone.js for.

And if so, for now if I target ES2022 and use async/await, I may want to ensure that even when I'm using signals or observables, that I should run detect changes myself after await to make sure I'm not leaving anything to chance?

And is it fair to say that because reactivity and the need for zone.js is being rethought and signals is a big part of this, that this situation will likely change?

**Comment by brentyates** [comment]
I have another update to report here... my initial finding of this situation wasn't using a timer as my repro examples use, it was awaiting a promise from a firestore document get via angular/fire. When I went back and used the same timer example from stackblitz (which we now know why it wasn't working there), it works locally in the same app. But, my firestore document get using await still doesn't do anything... and that is likely because a) it might just be hitting firestore's cache layer, meaning nothing zone.js has patched gets hit, or b) maybe in that code path it's using a web socket it already has open, I'm not sure.. not sure if zone.s has that patched or not. So I think I will just set the signal inside of zone.run or call cdr.detectChanges in this case.
